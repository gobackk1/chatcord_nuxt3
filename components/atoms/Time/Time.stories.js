import AtomTime from './Time.vue'

export default {
  title: 'atoms/Time',
  component: AtomTime,
  argTypes: {
    unixTime: {
      description:
        '表示する日時のUNIX時間。0未満、7258086000より大きい入力値の動作は保証しない。',
    },
    format: {
      description:
        '日時のフォーマットを指定する。https://momentjs.com/docs/#/parsing/string-format/',
    },
  },
  parameters: {
    docs: {
      description: {
        component:
          'UNIX時間をヒューマンリーダブルな時間形式へ変換するコンポーネント。\nMoment.jsを利用している。https://momentjs.com/',
      },
    },
  },
}

const Template = (args) => ({
  components: { AtomTime },
  setup() {
    return { args }
  },
  template: '<atom-time :unix-time="args.unixTime" :format="args.format" />',
})

export const Default = Template.bind({})
Default.args = {
  unixTime: 1675177200,
}
Default.parameters = {
  docs: {
    source: {
      code: '<atom-time :unix-time="1675177200" />',
    },
  },
}

export const WithFormat = Template.bind({})
WithFormat.args = {
  unixTime: 1675177200,
  format: 'MM月DD日(ddd) HH:mm',
}
WithFormat.parameters = {
  docs: {
    source: {
      code: '<atom-time :unix-time="1675177200" format="MM月DD日(ddd) HH:mm" />',
    },
  },
}

export const WithValidationError = Template.bind({})
WithValidationError.args = {
  unixTime: 999999999999999,
}
WithValidationError.parameters = {
  docs: {
    description: {
      story:
        '大きすぎる値を入力すると、invalid dateと表示される。props validatorの返り値によってはコンソールにエラーが表示される。',
    },
    source: {
      code: '<cc-time :unix-time="999999999999999" />',
    },
  },
}
