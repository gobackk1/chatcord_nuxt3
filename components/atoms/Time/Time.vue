<template>
  <time v-if="check" :datetime="datetime">{{ formattedTime }} </time>
</template>

<script lang="ts">
import { defineComponent } from 'vue'
import moment from 'moment'

const MACHINE_READABLE_DATE_FORMAT = 'YYYY-MM-DDTHH:mm'

export default defineComponent({
  name: 'AtomsTime',
  props: {
    format: {
      type: String,
      default: 'YYYY年MM月DD日(ddd) HH:mm',
    },
    check: {
      type: Boolean,
      default: true,
    },
    unixTime: {
      type: Number,
      required: true,
      validator(value: number) {
        /**
         * NOTE: 1970年01月01日(木) 09:00 から 2200年01月01日(水) 00:00 を入力の境界値とする。
         * momentは極めて大きい・小さいUNIX時間（9999999999999など）を入力すると「Invalid date」を出力する。
         * つまり、validation errorが発生した場合は、このコンポーネントがHTMLの仕様に準拠しない可能性があることを意味する
         */
        if (value >= 0 && value <= 7258086000) {
          return true
        } else {
          return false
        }
      },
    },
  },
  computed: {
    datetime(): string {
      /**
       * NOTE: テキストノードで（曜日）などの日本文化圏独特の表記をするために、
       * datetime属性値はmachine-readableな形式で入力する
       */
      return moment.unix(this.unixTime).format(MACHINE_READABLE_DATE_FORMAT)
    },
    formattedTime(): string {
      return moment.unix(this.unixTime).format(this.format)
    },
  },
})
</script>
