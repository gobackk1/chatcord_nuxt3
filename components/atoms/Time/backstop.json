{
  "scenarios": [
    {
      "label": "Default",
      "url": "http://localhost:6006/iframe.html?args=&id=atoms-time--default&viewMode=story",
      "misMatchThreshold" : 0.1
    },
    {
      "label": "WithFormat",
      "url": "http://localhost:6006/iframe.html?args=&id=atoms-time--with-format&viewMode=story",
      "misMatchThreshold" : 0.1
    },
    {
      "label": "WithValidationError",
      "url": "http://localhost:6006/iframe.html?args=&id=atoms-time--with-validation-error&viewMode=story",
      "misMatchThreshold" : 0.1
    }
  ]
}