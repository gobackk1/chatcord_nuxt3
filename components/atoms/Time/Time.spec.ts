import { shallowMount } from '@vue/test-utils'
import Time from './Time.vue'

describe('_Cc_Timeコンポーネントのテスト', () => {
  beforeAll(() => {
    vi.mock('moment', async () => {
      const moment = await vi.importActual('moment')
      return moment
    })
  })
  describe('ViewModelのテスト', () => {
    describe('ブラックボックステスト', () => {
      describe('同値・境界値テスト', () => {
        describe('unixTime prop', () => {
          describe('有効同値', () => {
            it('1675177200を入力した場合「2023年02月01日(水) 00:00」と表示すること', async () => {
              const wrapper = shallowMount(Time, {
                props: {
                  unixTime: 1675177200,
                },
              })
              expect(wrapper.html()).toContain('2023年02月01日(水) 00:00')
              expect(wrapper.html()).toMatchSnapshot()
            })
          })
          describe.skip('無効同値', () => {
            // NOTE: 無効同値は境界値テストとTypescriptの静的解析で担保とする
          })
          describe('境界値のテスト', () => {
            it('0を入力した場合「1970年01月01日(木) 09:00」と表示すること', () => {
              const wrapper = shallowMount(Time, {
                props: {
                  unixTime: 0,
                },
              })
              expect(wrapper.html()).toContain('1970年01月01日(木) 09:00')
              expect(wrapper.html()).toMatchSnapshot()
            })
            it('-1を入力した場合、コンソールエラーが発生すること', () => {
              const consoleSpy = vi
                .spyOn(console, 'warn')
                .mockImplementation(() => ({}))
              shallowMount(Time, {
                props: {
                  unixTime: -1,
                },
              })
              expect(consoleSpy).toHaveBeenCalled()
            })
            it('7258086000を入力した場合「2200年01月01日(水) 00:00」と表示すること', () => {
              const wrapper = shallowMount(Time, {
                props: {
                  unixTime: 7258086000,
                },
              })
              expect(wrapper.html()).toContain('2200年01月01日(水) 00:00')
              expect(wrapper.html()).toMatchSnapshot()
            })
            it('7258086001を入力した場合コンソールエラーが発生すること', () => {
              const consoleSpy = vi
                .spyOn(console, 'warn')
                .mockImplementation(() => ({}))
              shallowMount(Time, {
                props: {
                  unixTime: 7258086001,
                },
              })
              expect(consoleSpy).toHaveBeenCalled()
            })
          })
        })
        describe('formattedTime prop', () => {
          describe('有効同値', () => {
            // NOTE: 有効同値をうまく整理できないかつAPIはライブラリ側でテストすると考えられるので
            // プロダクト側で良く使うフォーマットを有効同値としてテストする
            it('MM月DD日(ddd) HH:mm を入力した場合「02月01日(水) 00:00」と表示すること', () => {
              const wrapper = shallowMount(Time, {
                props: {
                  unixTime: 1675177200,
                  format: 'MM月DD日(ddd) HH:mm',
                },
              })
              expect(wrapper.html()).toContain('02月01日(水) 00:00')
              expect(wrapper.html()).toMatchSnapshot()
            })
          })
          describe.skip('無効同値', () => {
            // NOTE: 無効同値はライブラリ側テストとTypescriptの静的解析で担保とする
          })
        })
      })
      describe.skip('組み合わせテスト', () => {
        // NOTE: propsの組み合わせが少なく、最低限の組み合わせは同値・境界値でテストしているので、組み合わせテストはしない
      })
    })
  })
  describe('Viewのテスト', () => {
    describe.skip('UIインタラクションテスト', () => {
      // NOTE: このコンポーネントはUIインタラクションを持たないのでテストしない
    })
    describe.skip('探索的テスト', () => {
      // TODO: 探索的テスト方針を定める
    })
  })
})
