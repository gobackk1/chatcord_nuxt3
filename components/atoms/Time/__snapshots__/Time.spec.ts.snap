// Vitest Snapshot v1

exports[`_Cc_Timeコンポーネントのテスト > ViewModelのテスト > ブラックボックステスト > 同値・境界値テスト > formattedTime prop > 有効同値 > MM月DD日(ddd) HH:mm を入力した場合「02月01日(水) 00:00」と表示すること 1`] = `
<time
  datetime="2023-02-01T00:00"
>
  02月01日(水) 00:00
</time>
`;

exports[`_Cc_Timeコンポーネントのテスト > ViewModelのテスト > ブラックボックステスト > 同値・境界値テスト > unixTime prop > 境界値のテスト > 0を入力した場合「1970年01月01日(木) 09:00」と表示すること 1`] = `
<time
  datetime="1970-01-01T09:00"
>
  1970年01月01日(木) 09:00
</time>
`;

exports[`_Cc_Timeコンポーネントのテスト > ViewModelのテスト > ブラックボックステスト > 同値・境界値テスト > unixTime prop > 境界値のテスト > 7258086000を入力した場合「2200年01月01日(水) 00:00」と表示すること 1`] = `
<time
  datetime="2200-01-01T00:00"
>
  2200年01月01日(水) 00:00
</time>
`;

exports[`_Cc_Timeコンポーネントのテスト > ViewModelのテスト > ブラックボックステスト > 同値・境界値テスト > unixTime prop > 有効同値 > 1675177200を入力した場合「2023年02月01日(水) 00:00」と表示すること 1`] = `
<time
  datetime="2023-02-01T00:00"
>
  2023年02月01日(水) 00:00
</time>
`;
